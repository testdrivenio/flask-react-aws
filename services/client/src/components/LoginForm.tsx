import React from "react";
import { Formik, Form, Field } from "formik";
import { z } from "zod";
import {
  Box,
  Button,
  FormControl,
  FormLabel,
  Input,
  FormErrorMessage,
  VStack,
  Heading,
} from "@chakra-ui/react";
import { Navigate } from "react-router-dom";

const validationSchema = z.object({
  email: z.string().email("Enter a valid email"),
  password: z.string().min(1, "Password is required"),
});

// eslint-disable-next-line @typescript-eslint/no-unused-vars
type FormValues = z.infer<typeof validationSchema>;

interface LoginFormProps {
  onSubmit: (values: { email: string; password: string }) => Promise<void>;
  isAuthenticated: () => boolean;
}

const LoginForm: React.FC<LoginFormProps> = ({ onSubmit, isAuthenticated }) => {
  if (isAuthenticated()) {
    return <Navigate to="/" replace />;
  }
  return (
    <Box maxWidth="400px" margin="100px auto 0">
      <Heading as="h1" size="xl" textAlign="center" mb={6}>
        Log In
      </Heading>
      <Formik
        initialValues={{
          email: "",
          password: "",
        }}
        onSubmit={(values, { setSubmitting, resetForm }) => {
          onSubmit(values);
          resetForm();
          setSubmitting(false);
        }}
        validate={(values) => {
          try {
            validationSchema.parse(values);
            return {};
          } catch (error) {
            return (error as z.ZodError).formErrors.fieldErrors;
          }
        }}
      >
        {({ errors, touched, isSubmitting }) => (
          <Form>
            <VStack spacing={4}>
              <Field name="email">
                {({ field }) => (
                  <FormControl isInvalid={!!(errors.email && touched.email)}>
                    <FormLabel htmlFor="email">Email</FormLabel>
                    <Input
                      {...field}
                      id="email"
                      type="email"
                      placeholder="Enter an email address"
                    />
                    <FormErrorMessage>{errors.email}</FormErrorMessage>
                  </FormControl>
                )}
              </Field>
              <Field name="password">
                {({ field }) => (
                  <FormControl
                    isInvalid={!!(errors.password && touched.password)}
                  >
                    <FormLabel htmlFor="password">Password</FormLabel>
                    <Input
                      {...field}
                      id="password"
                      type="password"
                      placeholder="Enter a password"
                    />
                    <FormErrorMessage>{errors.password}</FormErrorMessage>
                  </FormControl>
                )}
              </Field>
              <Button
                type="submit"
                colorScheme="blue"
                width="full"
                isLoading={isSubmitting}
                bg="green.400"
                _hover={{ bg: "green.500" }}
              >
                Log In
              </Button>
            </VStack>
          </Form>
        )}
      </Formik>
    </Box>
  );
};

export default LoginForm;
