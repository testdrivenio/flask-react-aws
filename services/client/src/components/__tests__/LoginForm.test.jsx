import React from 'react';
import { cleanup } from '@testing-library/react';

import LoginForm from '../LoginForm';

afterEach(cleanup);

const props = {
  handleLoginFormSubmit: () => { return true },
  isAuthenticated: () => { return false },
}

it('renders properly', () => {
  const { getByText } = renderWithRouter(<LoginForm {...props} />);
  expect(getByText('Log In')).toHaveClass('title');
});

it('renders with default props', () => {
  const { getByLabelText, getByText } = renderWithRouter(<LoginForm {...props} />);

  const emailInput = getByLabelText('Email');
  expect(emailInput).toHaveAttribute('type', 'email');
  expect(emailInput).not.toHaveValue();

  const passwordInput = getByLabelText('Password');
  expect(passwordInput).toHaveAttribute('type', 'password');
  expect(passwordInput).not.toHaveValue();

  const buttonInput = getByText('Submit');
  expect(buttonInput).toHaveValue('Submit');
});

it("renders", () => {
  const { asFragment } = renderWithRouter(<LoginForm {...props} />);
  expect(asFragment()).toMatchSnapshot();
});
