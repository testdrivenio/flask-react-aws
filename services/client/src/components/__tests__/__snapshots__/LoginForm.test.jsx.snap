// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="title is-1"
    >
      Log In
    </h1>
    <hr />
    <br />
    <form>
      <div
        class="field"
      >
        <label
          class="label"
          for="input-email"
        >
          Email
        </label>
        <input
          class="input"
          id="input-email"
          name="email"
          placeholder="Enter an email address"
          type="email"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="input-password"
        >
          Password
        </label>
        <input
          class="input"
          id="input-password"
          name="password"
          placeholder="Enter a password"
          type="password"
          value=""
        />
      </div>
      <input
        class="button is-primary"
        type="submit"
        value="Submit"
      />
    </form>
  </div>
</DocumentFragment>
`;
